File: typescript/debugger/debugger_test.ts
// Define types/interfaces to match Python's dataclasses
interface Product {
  id: number;
  name: string;
  price: number;
  category: string;
}

interface InvoiceSummary {
  itemsProcessed: number;
  subtotal: number;
  finalTotal: number;
  categories: Set<string>;
}

function applyDiscount(price: number, discountRate: number): number {
  // Calculates the discounted price. Good for testing 'step' and 'return'.
  const reduction = price * discountRate;
  const finalPrice = price - reduction;
  return Math.round(finalPrice * 100) / 100; // JavaScript/TypeScript rounding to 2 decimal places
}

function generateInvoice(
  inventory: Product[],
  discountMapping: Record<string, number>,
): InvoiceSummary {
  // Processes products and aggregates totals.
  const invoiceSummary: InvoiceSummary = {
    itemsProcessed: 0,
    subtotal: 0.0,
    finalTotal: 0.0,
    categories: new Set<string>(),
  };

  for (const item of inventory) {
    // 💡 Try setting a conditional breakpoint here: item.category === 'Electronics'
    invoiceSummary.itemsProcessed += 1;
    invoiceSummary.subtotal += item.price;
    invoiceSummary.categories.add(item.category);

    // Determine if a discount applies to this category (fallback to 0.0 if undefined)
    const rate = discountMapping[item.category] ?? 0.0;

    // 💡 Step 'into' this function or step 'over' it in your IDE debugger
    const itemFinalPrice = applyDiscount(item.price, rate);
    invoiceSummary.finalTotal += itemFinalPrice;
  }

  invoiceSummary.subtotal = Math.round(invoiceSummary.subtotal * 100) / 100;
  invoiceSummary.finalTotal = Math.round(invoiceSummary.finalTotal * 100) / 100;

  return invoiceSummary;
}

// Main execution block
function main() {
  debugger;

  // Sample catalog data
  const catalog: Product[] = [
    { id: 101, name: "Wireless Mouse", price: 29.99, category: "Electronics" },
    {
      id: 102,
      name: "Mechanical Keyboard",
      price: 89.99,
      category: "Electronics",
    },
    { id: 103, name: "Running Shoes", price: 120.0, category: "Apparel" },
    { id: 104, name: "Coffee Mug", price: 15.5, category: "Kitchen" },
  ];

  // Category discount rates (e.g., 10% off electronics, 20% off apparel)
  const discounts: Record<string, number> = {
    Electronics: 0.1,
    Apparel: 0.2,
  };

  console.log("Initializing invoice generation...");

  // 💡 Execution stops here if stepping through
  const report = generateInvoice(catalog, discounts);

  console.log("\n--- Final Invoice Report ---");
  // 💡 Use console.log(report) or console.table(report) to see a beautifully formatted output
  console.log(`Total Items: ${report.itemsProcessed}`);
  console.log(`Subtotal:    $${report.subtotal}`);
  console.log(`Total Due:   $${report.finalTotal}`);
}

// Run the main function
main();

File: python/debugger/debugger_test.py
from dataclasses import dataclass
from typing import List, Dict


@dataclass
class Product:
    id: int
    name: str
    price: float
    category: str


@dataclass
class InvoiceSummary:
    items_processed: int
    subtotal: float
    final_total: float
    categories: set[str]


def apply_discount(price: float, discount_rate: float) -> float:
    """Calculates the discounted price. Good for testing 'step' (s) and 'return' (r)."""
    reduction = price * discount_rate
    final_price = price - reduction
    return round(final_price, 2)


def generate_invoice(
    inventory: List[Product], discount_mapping: Dict[str, float]
) -> InvoiceSummary:
    """Processes products and aggregates totals. Good for testing loops and local variables."""
    invoice_summary = InvoiceSummary(
        items_processed=0, subtotal=0.0, final_total=0.0, categories=set()
    )

    for item in inventory:
        # 💡 Try setting a conditional breakpoint here: b 31, item.category == 'Electronics'
        invoice_summary.items_processed += 1
        invoice_summary.subtotal += item.price
        invoice_summary.categories.add(item.category)

        # Determine if a discount applies to this category
        rate = discount_mapping.get(item.category, 0.0)

        # 💡 Step 'into' this function using 's', or step 'over' it using 'n'
        item_final_price = apply_discount(item.price, rate)
        invoice_summary.final_total += item_final_price

    invoice_summary.subtotal = round(invoice_summary.subtotal, 2)
    invoice_summary.final_total = round(invoice_summary.final_total, 2)

    return invoice_summary


if __name__ == "__main__":
    breakpoint()

    # Sample catalog data
    catalog = [
        Product(id=101, name="Wireless Mouse", price=29.99, category="Electronics"),
        Product(
            id=102, name="Mechanical Keyboard", price=89.99, category="Electronics"
        ),
        Product(id=103, name="Running Shoes", price=120.00, category="Apparel"),
        Product(id=104, name="Coffee Mug", price=15.50, category="Kitchen"),
    ]

    # Category discount rates (e.g., 10% off electronics, 20% off apparel)
    discounts = {"Electronics": 0.10, "Apparel": 0.20}

    print("Initializing invoice generation...")

    # 💡 Execution stops here if you set a breakpoint at the start of main
    report = generate_invoice(catalog, discounts)

    print("\n--- Final Invoice Report ---")
    # 💡 Use 'pp report' here to see a beautifully formatted dictionary output
    print(f"Total Items: {report.items_processed}")
    print(f"Subtotal:   ${report.subtotal}")
    print(f"Total Due:  ${report.final_total}")
